klippy/extras/ads1x1x.py

```diff
@@ -0,0 +1,358 @@
+        buff = bytearray(params['response'])
+        return (buff[0]<<8 | buff[1])
+
+    def _write_register(self, reg, data):
+        data = [
+            (reg & 0xFF), # Control register
+            ((data>>8) & 0xFF), # High byte
+            (data & 0xFF), # Lo byte
+        ]
+        self._i2c.i2c_write(data)
+
+    def reset_all_devices(self):
+        try:
+            # Init all devices on bus for this kind of device
+            self._i2c.i2c_write([0x06, 0x00, 0x00])
+        except Exception:
+            logging.exception("ADS1X1X: error while resetting device")
+
+class ADS1X1X_pin:
+    def __init__(self, chip, pin_params):
+        self.mcu = chip.mcu
+        self.chip = chip
+        self.pin = pin_params['pin']
+
+        if (self.pin not in ADS1X1X_MUX):
+            raise pins.error('ADS1x1x pin %s is not valid' % self.pin)
+
+        self.mux = ADS1X1X_MUX[self.pin]
+        # Set up 2-byte configuration that will be used with each request
+        self.config = 0
+        self.config |= (ADS1X1X_OS['OS_SINGLE'] \
+            & ADS1X1X_REG_CONFIG['OS_MASK'])
+        self.config |= (self.mux & ADS1X1X_REG_CONFIG['MULTIPLEXER_MASK'])
+        self.config |= (chip.pga & ADS1X1X_REG_CONFIG['PGA_MASK'])
+        self.config |= (chip.mode & ADS1X1X_REG_CONFIG['MODE_MASK'])
+        self.config |= (chip.samples_per_second & \
+                        ADS1X1X_REG_CONFIG['DATA_RATE_MASK'])
+        self.config |= (chip.comp_mode \
+            & ADS1X1X_REG_CONFIG['COMPARATOR_MODE_MASK'])
+        self.config |= (chip.comp_polarity \
+            & ADS1X1X_REG_CONFIG['COMPARATOR_POLARITY_MASK'])
+        self.config |= (chip.comp_latching \
+            & ADS1X1X_REG_CONFIG['COMPARATOR_LATCHING_MASK'])
+        self.config |= (chip.comp_queue \
+            & ADS1X1X_REG_CONFIG['COMPARATOR_QUEUE_MASK'])
+
+        self.report_time = 1.0 / chip.samples_per_second_numeric * 4
+        self.invalid_count = 0
+
+        self.chip._printer.register_event_handler("klippy:connect", \
+                                            self._handle_connect)
+
+    def _handle_connect(self):
+        self._reactor = self.chip._printer.get_reactor()
+        self._sample_timer = \
+                self._reactor.register_timer(self._timer)
+        self._reactor.update_timer(self._sample_timer, self._reactor.NOW)
+
+    def _timer(self, eventtime):
+        sample = self.chip.sample(self)
+        if sample is not None:
+            self._process_sample(sample)
+
+        return self._reactor.monotonic() + self.report_time
+
+    def _process_sample(self, sample):
+        # The sample is encoded in the top 12 or full 16 bits
+        # Value's meaning is defined by ADS1X1X_REG_CONFIG['PGA_MASK']
+        if isADS101X(self.chip.chip):
+            sample >>= 4
+            target_value = sample / ADS101X_RESOLUTION
+        else:
+            target_value = sample / ADS111X_RESOLUTION
+
+        if self.maxval > self.minval:
```

Comment:
The calling code should verify `maxval > minval` and always provide values - so this looks like an odd check.

Reply:
Ah, I wasn't sure if it does for certain, so I just added it to be safe. I'll remove that part.